backend/danswer/connectors/teams/connector.py

```diff
@@ -0,0 +1,288 @@
+            )
+
+            if start and message_datetime < start:
+                continue
+            if end and message_datetime > end:
+                continue
+
+            reply_query = base_message.replies.get_all()
+            replies = reply_query.execute_query()
+
+            # start a list containing the base message and its replies
+            thread: list[ChatMessage] = [base_message]
+            thread.extend(replies)
+
+            threads.append(thread)
+
+        return threads
+
+    def _get_channels_from_teams(
+        self,
+        teams: list[Team],
+    ) -> list[Channel]:
+        channels_list: list[Channel] = []
+        for team in teams:
+            query = team.channels.get()
+            channels = query.execute_query()
+            channels_list.extend(channels)
+
+        return channels_list
+
+    def _get_all_teams(self) -> list[Team]:
+        if self.graph_client is None:
+            raise ConnectorMissingCredentialError("Teams")
+
+        teams_list: list[Team] = []
+
+        teams = self.graph_client.teams.get().execute_query()
+
+        if len(self.requested_team_list) > 0:
+            for requested_team in self.requested_team_list:
+                adjusted_request_string = requested_team.replace(" ", "")
+                for team in teams:
+                    adjusted_team_string = team.display_name.replace(" ", "")
+                    if adjusted_team_string == adjusted_request_string:
+                        teams_list.append(team)
+        else:
+            teams_list.extend(teams)
+
+        return teams_list
+
+    def _fetch_from_teams(
+        self, start: datetime | None = None, end: datetime | None = None
+    ) -> GenerateDocumentsOutput:
+        if self.graph_client is None:
+            raise ConnectorMissingCredentialError("Teams")
+
+        teams = self._get_all_teams()
+
+        channels = self._get_channels_from_teams(
+            teams=teams,
+        )
+
+        # goes over channels, converts them into Document objects and then yields them in batches
+        doc_batch: list[Document] = []
+        for channel in channels:
+            thread_list = self._get_threads_from_channel(channel, start=start, end=end)
+            for thread in thread_list:
+                converted_doc = self._convert_thread_to_document(channel, thread)
+                if converted_doc:
+                    doc_batch.append(converted_doc)
+
+            if len(doc_batch) >= self.batch_size:
+                yield doc_batch
+                doc_batch = []
+        yield doc_batch
+
+    def _construct_semantic_identifier(
+        self, channel: Channel, top_message: ChatMessage
+    ) -> str | None:
+        first_poster = (
+            top_message.properties.get("from", {})
+            .get("user", {})
+            .get("displayName", {})
+        )
+        channel_name = channel.properties["displayName"]
```

Comment:
Decision here between letting it break to catch errors and get a better understanding or also use a get since there's  not much to be done if it doesn't exist anyway.